src/project_root.rs

```diff
@@ -1,7 +1,18 @@
 use glob::glob;
 use std::path::PathBuf;
+use std::collections::HashSet;
+use std::env;
 
-pub fn find_project_root(roots: &[String], path: &str) -> String {
+pub fn find_project_root(language: &str, markers: &[String], path: &str) -> String {
+    let vars = gather_env_roots(language);
+    if vars.is_empty() {
+    	roots_by_marker(markers, path)
+    } else {
```

Comment:
I think that it might be good to fall back to `root_by_marker` not only when no appropriate env vars are found, but also when `roots_by_env` didn't match any of vars. In that case, visiting random project while running the same kak-lsp session would give the expected result.